### PR TITLE
Use reference assemblies when compiling Razor views

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/dotnet/aspnetcore/issues/18323

### CODE REVIEW

**Comment by JunTaoLuo** [comment]
Please fill out the servicing template and wait for shiproom approval. We want to take this in the 3.1 patch.

**Review by JunTaoLuo — APPROVED** [review]
Changes make sense to me

**Review by rainersigwald — APPROVED** [review]

**Comment by wtgodbe** [comment]
@pranavkm looks like this broke other projects, maybe have the target conditionally only run for projects that set some property (specifically the inner builds of identity.ui)?

**Comment by dougbu** [comment]
Once we get the **Right Thing:tm:** in this PR (perhaps, as @wtgodbe suggested, by narrowing when it's applied) should we instead take this as a tell-mode change @Pilchie❔

**Comment by pranavkm** [comment]
We verified that this also affects all the other UI projects (Azure.UI etc). I'm going to play around with this to see what's causing the build failure.

**Comment by wtgodbe** [comment]
I think this should be fine for tell mode - it's treating the same symptom as John's multi-target PR, which was approved, and that arguably could've been infra tell mode as well. @Pilchie thoughts?

File: src/Components/test/testassets/TestServer/Components.TestServer.csproj
Comment (pranavkm) [inline src/Components/test/testassets/TestServer/Components.TestServer.csproj]: This one project had an intersection of Components from the shared fx and Components as a package via Blazor. Using the ref assembly to compile here caused version mismatch during compile. This seems to be the only project that's affected by the workaround.

**Comment by Pilchie** [comment]
Let’s use tell mode for this one. I’d rather get the build unblocked ASAP.

**Review by rynowak — APPROVED** [review]

**Pull request merged**
